src/elements/src/index.ts

```diff
@@ -0,0 +1,55 @@
+import { LifeCycles } from 'single-spa';
+import { NgElement } from '@angular/elements';
+
+import {
+  SingleSpaAngularElementsOptions,
+  BootstrappedSingleSpaAngularElementsOptions,
+} from './types';
+
+const defaultOptions: BootstrappedSingleSpaAngularElementsOptions = {
+  element: null!,
+  ngModuleRef: null,
+  bootstrapFunction: null!,
+  withProperties: undefined,
+  domElementGetter: undefined,
+};
+
+async function bootstrap(options: BootstrappedSingleSpaAngularElementsOptions) {
+  if (options.ngModuleRef !== null) {
+    return;
+  }
+
+  options.ngModuleRef = await options.bootstrapFunction();
+}
+
+async function mount(options: BootstrappedSingleSpaAngularElementsOptions) {
+  if (typeof options.withProperties === 'function') {
+    const properties = await options.withProperties(options.ngModuleRef!);
+    const element = document.querySelector(options.element) as NgElement;
+
+    for (const [property, value] of Object.entries(properties)) {
+      element[property] = value;
+    }
+  }
+}
+
+function unmount(options: BootstrappedSingleSpaAngularElementsOptions) {
```

Comment:
It's better to have the lifecycle function return a promise before throwing any errors, so that errors are a promise rejection instead of synchronous error. To do that, there are two options:

```js
async function unmount(options) {
  const node = document.querySelector(options.element)
  node.parentElement.removeChild()
}
```

or:

```js
function unmount(options) {
  return Promise.resolve().then(() => {
    const node = document.querySelector(options.element)
    node.parentElement.removeChild()
  })
}
```